blockprotocol/blockprotocol
# Implement block publishing flow

## 🌟 What is the purpose of this PR?

This adds `Publish a block` interface for logged-in users. So it will be easier to publish new blocks to `Block Hub`. For now, we only support `Publish from npm` method. So users can publish a valid block as a `npm package`, and add that block to `Block Hub` via `Publish from npm` method.

## 🔗 Related links

- [Asana task - publish block](https://app.asana.com/0/1202727124867670/1202796290587274/f) _(internal)_
- [Asana task - list published blocks](https://app.asana.com/0/0/1202796290587276/f) _(internal)_
- [Design](https://www.figma.com/file/Fpmde3E9nJMb4VLkPTYgrM/blockprotocol.org?node-id=3895%3A29099) _(internal)_

## 🔍 What does this change?

- add a wrapper for `MUI <Alert />`, and override theme options for `Alert` component according to design.
  - also add eslint rule to prevent importing `Alert` from `mui`
- add `react-hook-form` to project, for better form validation / state handling
- update`playwright` `1.24.0` -> `1.25.1`
  -  **Reason:** To have `Pick Selector` & `Record new tests from current page state` features on Playwright VSCode Extension. 
  - checked the  [1.25 Changelog](https://playwright.dev/docs/release-notes#version-125) to make sure there is no breaking change.
  - also tried running tests locally to make sure everything works as expected after the update. 
- disable the `Button` component when `loading={true}`
- Improved `ListViewCard` component, now it supports `extraContent`, and could be used with custom content. (see [pages/blocks/index.page.tsx](https://github.com/blockprotocol/blockprotocol/blob/a55bc705aa0eb82f31e33678d59608f0afc5c886/site/src/pages/blocks/index.page.tsx#L78-L91))
- create `formatUpdatedAt` helper function
- add `error?.code` to `ApiClientError`
- add `bpCode` typography variant for `code` style
- Add `green` & new purple variants to `palette`

## 🐾 Next steps

- Implement `SWR` for data fetching logic instead of manually handling following:
  - Setting data to state
  - Handling errors
  - Loading status

## 🛡 What tests cover this?

- Integration tests at `site/tests/integration/block-publish-npm.test.ts`

## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  Create a new use
2. Open dashboard
3. Select `Blocks` tab from above
4. Confirm the empty state of the blocks list
5. Publish a new block via npm. Here is a [npm package](https://www.npmjs.com/package/test-npm-block) for testing this functionality.
6. Get redirected back to `Blocks` page after publishing
7. Confirm the filled state of the blocks list

## 📹 Demo
<img width="1728" alt="1" src="https://user-images.githubusercontent.com/39553853/187151400-4de59464-201e-43a2-b281-70b8b9230dd3.png">
<img width="1728" alt="2" src="https://user-images.githubusercontent.com/39553853/187151420-79f7289d-1cf7-4a2d-bde8-2325783a6206.png">
<img width="1728" alt="3" src="https://user-images.githubusercontent.com/39553853/187151429-1bd3a862-206d-4ffc-8d9f-d3ac6cb95c13.png">
<img width="1728" alt="4" src="https://user-images.githubusercontent.com/39553853/187151434-40464be4-902b-4dc1-b4b4-2003bbcb871f.png">
<img width="1728" alt="5" src="https://user-images.githubusercontent.com/39553853/187151437-e6671d7c-d811-4c99-93b7-9e5d8d6abfdb.png">
